gateway/reverse_proxy.go

```diff
@@ -380,44 +384,56 @@ type ReverseProxy struct {
 	TykAPISpec   *APISpec
 	ErrorHandler ErrorHandler
 
+	// transportPool holds *http.Transports for reuse.
+	transportPool *httputil.TransportPool
+
 	logger *logrus.Entry
-	sp     sync.Pool
 	Gw     *Gateway `json:"-"`
 }
 
+// A BufferPool is an interface for getting and returning temporary
+// byte slices for use by io.CopyBuffer.
+type BufferPool interface {
+	Get() []byte
+	Put([]byte)
+}
+
 var idleConnTimeout = 90
 
-func (p *ReverseProxy) defaultTransport(dialerTimeout float64) *http.Transport {
-	timeout := 30.0
-	if dialerTimeout > 0 {
-		log.Debug("Setting timeout for outbound request to: ", dialerTimeout)
-		timeout = dialerTimeout
+func (p *ReverseProxy) defaultTransport(timeout int) (transport *http.Transport) {
+	dialerTimeout := 30
+	if timeout > 0 {
+		dialerTimeout = timeout
 	}
 
-	dialer := &net.Dialer{
-		Timeout:   time.Duration(float64(timeout) * float64(time.Second)),
-		KeepAlive: 30 * time.Second,
-		DualStack: true,
-	}
-	dialContextFunc := dialer.DialContext
-	if p.Gw.dnsCacheManager.IsCacheEnabled() {
-		dialContextFunc = p.Gw.dnsCacheManager.WrapDialer(dialer)
-	}
+	dialContextFunc := p.dialContextFunc(dialerTimeout)
+
+	gwConfig := p.Gw.GetConfig()
+
+	// Construct a key from values that apply to the transport.
+	key := fmt.Sprintf("%d;%d;%d", dialerTimeout, gwConfig.MaxIdleConns, gwConfig.MaxIdleConnsPerHost)
```

Comment:
As per discussion with @jeffy-mathew, using APISpec.ID makes more sense in here, to isolate transports better than just over a timeout value. Other fields inside transport may be configured differently (TLS, etc.).

Reply:
Using APIID now